checkbox-ng/checkbox_ng/launcher/checkbox_cli.py

```diff
@@ -58,45 +64,51 @@ def __init__(self, args, sa):
 
 
-def main():
-    import argparse
-
-    commands = {
-        "check-config": CheckConfig,
-        "launcher": Launcher,
-        "list": List,
-        "run": Run,
-        "startprovider": StartProvider,
-        "submit": Submit,
-        "show": Show,
-        "list-bootstrapped": ListBootstrapped,
-        "merge-reports": MergeReports,
-        "merge-submissions": MergeSubmissions,
-        "tp-export": TestPlanExport,
-        "service": RemoteSlave,
-        "remote": RemoteMaster,
-    }
-    deprecated_commands = {
-        "slave": "service",
-        "master": "remote",
-    }
-
-    known_cmds = list(commands.keys())
-    known_cmds += list(deprecated_commands.keys())
-    known_cmds += ["-h", "--help"]
-    if not (set(known_cmds) & set(sys.argv[1:])):
-        sys.argv.insert(1, "launcher")
-
-    for i, arg in enumerate(sys.argv):
-        if arg in deprecated_commands:
-            sys.argv[i] = deprecated_commands[arg]
-            logging.warning(
-                "%s is deprecated. Please use %s instead",
-                arg,
-                deprecated_commands[arg],
+def parse_args(parser, default_command, deprecated_commands):
+    sys_argv = copy(sys.argv)
+    try:
+        return parser.parse_args(args=sys_argv[1:])
+    except argparse.ArgumentError as e:
+        error = e
+    # this could be caused by:
+    #   usage of default_command
+    #   usage of a deprecated command
+    #   a typo
+    if error.argument_name == "subcommand":
+        # get all deprecated command used
+        deprecated_sys_argv = [arg for arg in sys_argv if arg in deprecated_commands]
+        if deprecated_sys_argv:
+            # -> usage of deprecated command
+            dep_command = deprecated_sys_argv[0]
+            new_command = deprecated_commands[dep_command]
+            _logger.warning(
+                "%s is deprecated. Please use %s instead!", dep_command, new_command
             )
+            # replace the first deprecated command (assuming any other is a file arg)
+            sys_argv[sys_argv.index(dep_command)] = new_command
+        else:
+            # -> usage of default_command
+            # discover where the args for the default_command start
+            i = len(sys_argv)
+            while i > 1:
+                try:
+                    parser.parse_args(args=sys_argv[1:i])
+                    break
+                except argparse.ArgumentError:
+                    i-=1
+            # apply the default command
+            sys_argv.insert(i, default_command)
+        with contextlib.suppress(argparse.ArgumentError):
+            return parser.parse_args(args=sys_argv[1:])
+    # -> typo is some arg
+    parser.print_usage()
+    raise SystemExit(str(error))
 
-    top_parser = argparse.ArgumentParser()
+
+def setup_and_parse_args(default_command, commands, deprecated_commands={}):
+    top_parser = argparse.ArgumentParser(
+        conflict_handler="resolve", exit_on_error=False
```

Comment:
[`exit_on_error`](https://docs.python.org/3/library/argparse.html#argumentparser-objects) is python3.9 only unfortunately

Reply:
Thank you for catching that, fixed!